### ISSUE TITLE
Bug: [no-non-null-asserted-optional-chain] Rule not working as expected if non-null-assertion is followed by other actions

### ISSUE THREAD

**Issue by tathagat2000** [typescript-eslint/typescript-eslint#8670]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.0.2&fileType=.ts&code=MYewdgzgLgBAHgLhgQzATxgXhgbwL4BQBA9MQCIgCmEMYIsA7iAE4DWKNlcADpcFJQAmBUJFgZscAPwA6ZLIBGAQhkBbZNwAUALywA%2BGNoCURUgHUWrGsk48%2BA4aOiGs8WfJnKgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUFvcgAQBcBPABxQGNoBLEggWhXioDsCB6Zgezq%2BZ4Xh0AhsmSJoBRABM6nWlU7Mhg8gAshLdFHHRO0SODABfEEaA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const x: any = {}

//Does not work as expected
const y = x?.a?.b!.map(z => z)

//Works as expected
const z = x?.a?.b!
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
        "@typescript-eslint/no-non-null-asserted-optional-chain": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

I expect that the below line should throw an error

`const y = x?.a?.b!.map(z => z)`

### Actual Result

No error was thrown for this line

`const y = x?.a?.b!.map(z => z)`

**Comment by bradzacher** [comment]
This is working as expected!

In the error case you have an expression that may return `undefined` - so by using a non-null assertion you are cancelling the undefined and creating a bug vector in the code.

However with the continued chain case the optional chain short-circuits - meaning you may never reach the non-null assertion at all.
Specifically when the non-null assertion is directly chained off it does not apply to the entire expression - instead it applies just to the property it is directly attached to.
I.e. In your "bad" case the non-null assertion is not removing the undefined from the optional chain - instead it's removing the nullishness from the `b` property.

This is obviously still a bug vector - non-nullish assertions are never truly safe - but it's not a "this is always completely wrong" usage like the former.

**Comment by Josh-Cena** [comment]
If you want to catch this case, use `no-unnecessary-type-assertion`
